java/java-analysis-impl/src/com/intellij/psi/impl/source/resolve/reference/impl/JavaReflectionReferenceUtil.java

```diff
@@ -487,28 +528,94 @@ public static ReflectiveSignature composeMethodSignature(@Nullable PsiExpression
       final PsiMethodCallExpression methodCallExpression = (PsiMethodCallExpression)typeDefinition;
       final String referenceName = methodCallExpression.getMethodExpression().getReferenceName();
 
-      Function<PsiExpression[], ReflectiveSignature> composer = null;
+      final PsiMethod method = methodCallExpression.resolveMethod();
```

Comment:
Likely, `CallMatcher.staticCall(JAVA_LANG_INVOKE_METHOD_TYPE, METHOD_TYPE, GENERIC_METHOD_TYPE)` matcher will work here.

Reply:
You can even create a dynamic signature-based mapper, like this (assuming `CallMatcher.*` static import):

```java
      private static final CallMapper<ReflectiveSignature> SIGNATURE_MAPPER = new CallMapper<ReflectiveSignature>()
        .register(anyOf(
                    staticCall(JAVA_LANG_INVOKE_METHOD_TYPE, METHOD_TYPE).parameterCount(1),
                    staticCall(JAVA_LANG_INVOKE_METHOD_TYPE, METHOD_TYPE).parameterCount(3),
                    staticCall(JAVA_LANG_INVOKE_METHOD_TYPE, METHOD_TYPE).parameterTypes(JAVA_LANG_CLASS, JAVA_LANG_CLASS)),
                  call -> composeMethodSignatureFromTypes(call.getArgumentList().getExpressions()))
        .register(staticCall(JAVA_LANG_INVOKE_METHOD_TYPE, GENERIC_METHOD_TYPE),
                  call -> composeGenericMethodSignature(call.getArgumentList().getExpressions()))
        .register(staticCall(JAVA_LANG_INVOKE_METHOD_TYPE, METHOD_TYPE).parameterTypes(JAVA_LANG_CLASS, JAVA_LANG_INVOKE_METHOD_TYPE),
                  call -> composeMethodSignatureFromReturnTypeAndMethodType(call.getArgumentList().getExpressions()))
        .register(staticCall(JAVA_LANG_INVOKE_METHOD_TYPE, METHOD_TYPE).parameterTypes(JAVA_LANG_CLASS, JAVA_LANG_CLASS + "[]"),
                  call -> composeMethodSignatureFromReturnTypeAndListOrArray(call.getArgumentList().getExpressions()))
        .register(staticCall(JAVA_LANG_INVOKE_METHOD_TYPE, METHOD_TYPE).parameterTypes(JAVA_LANG_CLASS, JAVA_UTIL_LIST),
                  call -> composeMethodSignatureFromReturnTypeAndListOrArray(call.getArgumentList().getExpressions()));
      ...
      return SIGNATURE_MAPPER.mapFirst(methodCallExpression);
```

Btw, it's easy to dispatch array and list cases to two separate methods, so probably it would be nice to split `composeMethodSignatureFromReturnTypeAndListOrArray`

Reply:
> You can even create a dynamic signature-based mapper, like this (assuming CallMatcher.* static import):

Replaced with dynamic `CallMatcher` in both `JavaReflectionReferenceUtil` and `JavaLangReflectHandleInvocationChecker`

> so probably it would be nice to split composeMethodSignatureFromReturnTypeAndListOrArray

Split